_posts/2023-11-20-unlocking-the-secrets-to-ingestion.md

```diff
@@ -0,0 +1,191 @@
+
+### Primary Shard Distribution
+
+We want to ensure we prioritize balancing our shards across nodes so the workload is evenly distributed. Without setting this we could end up with a situation like the one below where all our primary shards for end up on the same node. That would cause our Node 1 here to be completely saturated while the other nodes may be underutilized.
+
+![Primary Shard Distribution](assets/media/blog-images/2023-11-20-unlocking-the-secrets-to-ingestion/PrimaryShardDistribution.svg){: .img-fluid }
+
+```json
+PUT _cluster/settings
+{
+	"persistent": {
+		"cluster.rounting.allocation.balanace.prefer_primary": True
+	}
+}
+```
+
+Enabling primary shard distribution ensures primary shards are given the first preference when it comes to spreading them across nodes. When this is enabled replicas may be more likely to end up on the same node however this is okay in an ingestion heavy workload.
+
+![Primary Shard Distribution Enabled](assets/media/blog-images/2023-11-20-unlocking-the-secrets-to-ingestion/PrimaryShardDistributionEnabled.svg){: .img-fluid }
+
+### Segment Replication
+
+One setting that dramatically impact ingestion is segment replication. Out of the box OpenSearch uses document replication. In this strategy data is first sent to the primary shard which then ingests it back into a Lucene segment. After it’s done that it will send the original unprocessed document to any replicas. They will then ingest the document back to their own Lucene segments. This works really well if you need more immediate consistency however it is duplicating the work done on the CPU for each replica.
+
+![Document Replication](/assets/media/blog-images/2023-11-20-unlocking-the-secrets-to-ingestion/DocumentReplication.svg){: .img-fluid }
+
+With segment replication enabled the first two steps remain the same. Once it gets to the third step is where the magic happens. Instead of sending the original unprocessed document into the replicas we send the processed Lucene segment over the network. Now the nodes that the replicas live on will be free to use more of their compute for ingestion. This comes with a few caveats though. The first is you will be using more bandwidth between nodes as the segments are much larger than the initial documents. It’s important to consider your network topology here. Second, this model is an eventually consistent model as the document updates are not available as readily on replicas. The final, is you will want to increase the refresh interval of the nodes so you do not overwhelm the replica nodes with network traffic. We’ll discuss that more in the next section.
```

Comment:
```suggestion
With segment replication enabled the first two steps remain the same. Once it gets to the third step is where the magic happens. Instead of sending the original unprocessed document into the replicas, we send the processed Lucene segment over the network. Now the nodes that the replicas live on will be free to use more of their compute for ingestion. This comes with a few caveats though. The first is you will be using more bandwidth between nodes, as the segments are much larger than the initial documents. It’s important to consider your network topology here. Second, this model is an eventually consistent model, as the document updates are not available as readily on replicas. Last, you will want to increase the refresh interval of the nodes so that you do not overwhelm the replica nodes with network traffic. We’ll discuss that more in the next section.
```